### PR TITLE
Sync tmp directory

### PR DESCRIPTION
**Pull request by b-chu**
## Description of changes:
Use a temporary directory based on remote and split instead of a randomly created one so the same directory will be used in distributed training

## Issue #, if available:
#293 

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/base/stream.py [hunk 1]
```diff
@@ -145,6 +149,15 @@ def __init__(self,
             self.keep_zip = keep_zip
             self.safe_keep_zip = self.keep_zip or self.remote in {None, self.local}
 
+    def _get_temporary_directory(self) -> str:
+        root = tempfile.gettempdir()
```
Comment (karan6181) [inline streaming/base/stream.py:153]: What would be the value of `root` when running on 2 GPUs? Does each rank gives the same `root` directory ?

File: streaming/base/stream.py [hunk 2]
```diff
@@ -98,8 +99,11 @@ def __init__(self,
                  keep_zip: Optional[bool] = None) -> None:
         self.remote = remote
         self._local = local
-        self.local = local or mkdtemp()
         self.split = split or ''
+        if local is None:
```
Comment (karan6181) [inline streaming/base/stream.py:103]: What would happen if user instantiate two or more `StreamingDataset` with same `remote` and `local=None` ? Ideal behavior is to throw a valid error message asking user to provide a `local` directory.

File: streaming/base/stream.py [hunk 3]
```diff
@@ -145,6 +149,15 @@ def __init__(self,
             self.keep_zip = keep_zip
             self.safe_keep_zip = self.keep_zip or self.remote in {None, self.local}
 
+    def _get_temporary_directory(self) -> str:
+        root = tempfile.gettempdir()
+        hash = None
+        if self.remote is not None:
+            hashlib.md5(self.remote.encode('utf-8')).hexdigest()
```
Comment (karan6181) [inline streaming/base/stream.py:156]: What is the max character length output of `hashlib.md5` ?

File: streaming/base/stream.py (hunk 1 above)
Comment (b-chu) [reply streaming/base/stream.py:153]: Yep they use the same `root`. I've changed the tests to use 2 GPUs and they all pass, but I'm getting a weird error at the end of the pytest saying it didn't exit correctly

```
========================= 4 passed, 1 warning in 1.17s =========================

----------End global rank 1 STDOUT----------
----------Begin global rank 1 STDERR----------

----------End global rank 1 STDERR----------
ERROR:__main__:Global rank 0 (PID 6785) exited with code -15
```

File: streaming/base/stream.py (hunk 2 above)
Comment (b-chu) [reply streaming/base/stream.py:103]: With multiple `StreamingDataset` and multiple GPUs, how do we know if a directory is being created/accessed by the different GPUs within the same `StreamingDataset` or by different `StreamingDataset`s? Does streaming have functions similar to composer's [broadcast](https://github.com/mosaicml/composer/blob/dev/composer/utils/dist.py#L248C8-L248C8)? Should we use torch.dist?

File: streaming/base/stream.py (hunk 3 above)
Comment (b-chu) [reply streaming/base/stream.py:156]: 32

File: streaming/base/stream.py [hunk 4]
```diff
@@ -138,13 +140,54 @@ def __init__(self,
                 raise ValueError('`download_timeout` must be positive')
             self.download_timeout = download_timeout
 
+        self.local_directory_timeout = local_directory_timeout
+        if local_directory_timeout <= 0:
+            raise ValueError('`local_directory_timeout` must be positive')
+
         self.validate_hash = validate_hash
 
+        if local is None:
```
Comment (karan6181) [inline streaming/base/stream.py:149]: There are two cases where `torch.distributed.init_process_group()` is created for a distributed call. If it is initialized before creating the `StreamingDataset` then this solution would work. However, if it is not initialized then Streaming initialize it [here](https://github.com/mosaicml/streaming/blob/main/streaming/base/dataset.py#L322) which is after Streams are created. I would recommend to move the `destroy_dist = maybe_init_dist()` line before [this](https://github.com/mosaicml/streaming/blob/main/streaming/base/dataset.py#L290).

File: streaming/base/stream.py [hunk 5]
```diff
@@ -94,11 +96,11 @@ def __init__(self,
                  choose: Optional[int] = None,
                  download_retry: Optional[int] = None,
                  download_timeout: Optional[float] = None,
+                 local_directory_timeout: float = 60,
```
Comment (karan6181) [inline streaming/base/stream.py:99]: I would recommend not to add another parameter. Can we use `download_timeout` instead ? I don't think `mkdirs()` can take more than 1 sec, so we don't need to worry much about timeout but it's a good habit to raise an exception which you did.

File: streaming/base/stream.py [hunk 6]
```diff
@@ -138,13 +140,54 @@ def __init__(self,
...
 
+    def _get_temporary_directory(self) -> str:
+        """Construct a path to a temporary directory based on remote and split."""
+        root = tempfile.gettempdir()
+        hash = None
+        if self.remote is not None:
+            hash = hashlib.md5(self.remote.encode('utf-8')).hexdigest()
+        # Removes underscore if self.split is an empty string
+        folder = '_'.join(filter(None, [hash, self.split]))
+        return os.path.join(root, folder)
```
Comment (karan6181) [inline streaming/base/stream.py:168]: How about something like this ?

```python
import streaming.base.distributed as dist
local = os.path.join(root, folder)
if dist.get_local_rank() == 0:
    if os.path.exists(local):
            raise ValueError(f'Could not create a local directory. Specify a local directory with the `local` value.')
    os.makedirs(local)
dist.barrier()
```

File: streaming/base/stream.py [hunk 7]
```diff
@@ -138,13 +140,54 @@ def __init__(self,
...
 
+        self.local_directory_timeout = local_directory_timeout
+        if local_directory_timeout <= 0:
+            raise ValueError('`local_directory_timeout` must be positive')
+
         self.validate_hash = validate_hash
 
+        if local is None:
+            self.local = self._get_temporary_directory()
+            self._create_or_wait_for_local()
```
Comment (karan6181) [inline streaming/base/stream.py:151]: you don't need this method. Check my below comment

File: streaming/base/stream.py [hunk 8]
```diff
@@ -138,13 +140,54 @@ def __init__(self,
...
+        """Construct a path to a temporary directory based on remote and split."""
+        root = tempfile.gettempdir()
+        hash = None
+        if self.remote is not None:
+            hash = hashlib.md5(self.remote.encode('utf-8')).hexdigest()
+        # Removes underscore if self.split is an empty string
+        folder = '_'.join(filter(None, [hash, self.split]))
+        return os.path.join(root, folder)
+
+    def _create_or_wait_for_local(self) -> None:
```
Comment (karan6181) [inline streaming/base/stream.py:170]: You don't need this method based on my above suggestion in another comment !

File: tests/test_stream.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+from typing import Optional
+
+import pytest
+from _pytest.monkeypatch import MonkeyPatch
+
+from streaming import Stream
+from streaming.base.distributed import barrier
+
+
+@pytest.mark.world_size(2)
```
Comment (karan6181) [inline tests/test_stream.py:17]: I am not 100% sure that by setting `@pytest.mark.world_size(2)` would create 2 sub-processes. Can you check once by printing the `RANK`, `WORLD_SIZE`, and `LOCAL_WORLD_SIZE` ?

File: tests/test_stream.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
+# Copyright 2023 MosaicML Streaming authors
```
Comment (karan6181) [inline tests/test_stream.py:1]: Can you also add a unit test for the exception that it generates ?

**Review by karan6181 — COMMENTED** [review]
Please also try running the basic dataset iteration on more than 1 gpu (maybe 2, 4, or 8) multiple times with `local=None` to ensure it works as expected.

File: tests/test_stream.py (hunk 9 above)
Comment (b-chu) [reply tests/test_stream.py:17]: Yep, shows the correct world size.

File: streaming/base/stream.py [hunk 11]
```diff
@@ -138,13 +140,54 @@ def __init__(self,
...
         if keep_zip is not None:
             self.keep_zip = keep_zip
             self.safe_keep_zip = self.keep_zip or self.remote in {None, self.local}
 
+    def _get_temporary_directory(self) -> str:
+        """Construct a path to a temporary directory based on remote and split."""
+        root = tempfile.gettempdir()
+        hash = None
+        if self.remote is not None:
+            hash = hashlib.md5(self.remote.encode('utf-8')).hexdigest()
```
Comment (knighton) [inline streaming/base/stream.py:165]: we do not want to be seen in public using md5, please use something like `sha3_256`

Comment (knighton) [reply streaming/base/stream.py:165]: probably want to truncate it for length though (say 16 bytes?)

File: streaming/base/stream.py [hunk 12]
```diff
@@ -138,13 +140,54 @@ def __init__(self,
...
+        # Removes underscore if self.split is an empty string
+        folder = '_'.join(filter(None, [hash, self.split]))
```
Comment (knighton) [inline streaming/base/stream.py:167]: need it to be like

```
root/
    hash/
        split/
```

so just `return os.path.join(root, hash, split)` i presume

File: streaming/base/dataset.py [hunk 13]
```diff
@@ -287,6 +288,9 @@ def __init__(self,
             raise ValueError(
                 'You must provide either `streams` or `remote`/`local`, but not both.')
 
+        # Create a temporary directory for the default stream
+        if remote is None and local is None and split is None:
+            local = mkdtemp()
```
[the remark this answers is not in the document]
Comment (knighton) [reply streaming/base/dataset.py:293]: and by `remote`/`local`, we mean our API here is that at least one of them must be set

`mkdtemp` relies on a seed not internal to streaming so is not really usable

[the remark this answers is not in the document]
Comment (b-chu) [reply streaming/base/dataset.py:293]: This is added because in the code path where `streams` is provided, a default stream can get created with remote/local/split all as None which raises ValueError(f'Could not create a local directory. Specify a local directory with the `local` value.') in stream.py. We need a valid temporary directory and that directory doesn't get used so it should be fine.

File: streaming/base/stream.py [hunk 14]
```diff
@@ -140,11 +141,31 @@ def __init__(self,
 
         self.validate_hash = validate_hash
 
+        if local is None:
+            self.local = self._get_temporary_directory()
+            if get_local_rank() == 0:
```
Comment (knighton) [inline streaming/base/stream.py:146]: why is this here? if i am not mistaken, `self.local` will get created incidentally when local rank zero downloads its index

File: streaming/base/dataset.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (knighton) [reply streaming/base/dataset.py:296]: lol that is gnarly

what do you think about passing `default` as a dict instead of instantiating a "default" stream?

File: streaming/base/stream.py (hunk 14 above)
Comment (b-chu) [reply streaming/base/stream.py:146]: I think we want to make sure that if multiple streaming datasets are created, they use different local directories instead of all being instantiated with None so we require that the temporary directory doesn't exist beforehand.

File: streaming/base/dataset.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (b-chu) [reply streaming/base/dataset.py:296]: Makes sense

**Review by karan6181 — APPROVED** [review]
